miniapps/autodiff/admfem.hpp

```diff
@@ -0,0 +1,674 @@
+                aduu[ii].setGradient(1.0);
+                tf(vparam,aduu,rr);
+                for(int jj=0;jj<vector_size;jj++)
+                {
+                    jac(jj,ii)=rr[jj].getGradient();
+                }
+                aduu[ii].setGradient(0.0);
+            }
+        }
+#else //end MFEM_USE_ADFORWARD
+        // use reverse mode
+        TFunctor<ad::ADFloatType, const Vector, ad::ADVectorType,
+                vector_size, state_size, param_size> tf;
+
+        jac.SetSize(vector_size, state_size);
+        jac = 0.0;
+        {
+            ad::ADVectorType aduu(state_size);
+            ad::ADVectorType rr(vector_size);
+            for(int i=0;i<state_size;i++){
+                aduu[i]=uu[i];
+            }
+
+            ad::ADFloatType::TapeType& tape =ad::ADFloatType::getGlobalTape();
+            typename ad::ADFloatType::TapeType::Position pos=tape.getPosition();
+
+            tape.setActive();
+            for(int ii=0;ii<state_size;ii++){ tape.registerInput(aduu[ii]); }
+            tf(vparam,aduu,rr);
+            for(int ii=0;ii<vector_size;ii++){ tape.registerOutput(rr[ii]); }
+            tape.setPassive();
+            for(int jj=0;jj<vector_size;jj++){
+                rr[jj].setGradient(1.0);
+                tape.evaluate();
+                for(int ii=0;ii<state_size;ii++){
+                    jac(jj,ii)=aduu[ii].getGradient();
+                }
+                tape.clearAdjoints();
+                rr[jj].setGradient(0.0);
+            }
+            tape.reset(pos);
+        }
+#endif
+    }
+};
+
+/// The class provides an evaluation of the first derivatives
+/// and the Hessian of a templated scalar function provided as
+///  a functor TFunctor. Both the first and the second derivatives
+/// are evaluated with the help of automatic differentiation (AD)
+/// https://en.wikipedia.org/wiki/Automatic_differentiation.
+/// The template parameters specify the size of the input
+/// vector (state_size) and the size of the parameters
+/// supplied to the function. The TFunctor functor is a template
+/// class with parameters [Float data type],
+/// [Vector type for the additional parameters],
+/// [Vector type for the state vector and the return residual].
+/// The integer template parameters are the same ones passed
+/// to QFunctionAutoDiff.
+template<template<typename, typename, typename, int, int> class TFunctor
+         , int state_size=1, int param_size=0>
+class QFunctionAutoDiff
+{
+public:
+
+    /// Evaluates a function for arguments vparam and uu.
+    /// The evaluation is based on the operator() in the
+    /// user provided functor TFunctor.
+    double QEval(const mfem::Vector &vparam, mfem::Vector &uu)
+    {
+        TFunctor<double, const mfem::Vector, mfem::Vector, state_size, param_size> tf;
+        return tf(vparam,uu);
+    }
+
+    /// Provides the same functionality as QGrad.
+    void QVectorFunc(const mfem::Vector &vparam, mfem::Vector &uu, mfem::Vector &rr)
```

Comment:
The method names in this class are confusing because it's not apparent to me why there are so many public methods, and I don't understand the motivation for the forwarding method names -- specifically why `QVectorFunc` forwards to `QGrad` and `QJacobian` forwards to `QHessian`, or what `Jacobian` is referring to in this context. It seems like `Grad` refers to the gradient of the function represented by `QEval` with respect to `uu`. But clearly `Jacobian` doesn't refer to the Jacobian of that function, because that quantity would be the adjoint of the gradient, not the Hessian of that function. What am I missing?

Reply:
The issue with the naming comes from two cases below: 

First case) The user starts with a scalar function. The gradients will be the first derivative, and the Hessian will hold the second derivatives. 

Second case) The user starts with a vector function, and then the Jacobian is computed on this vector function. 


Now let say a user starts implementation with a scalar function representing some kind of energy. The gradients will add to the residual, and they will be computed by AD. Next, the Hessian will add to the tangent matrix, and it will be computed by AD again.  The code will go through testing, and the user is happy with the implementation.  At his point, he/she will want to improve performance. The most straightforward modification would be implementing the first derivative as a vector function and applying AD to compute its Jacobian. The main code will not require any modifications - by having name duplication of Jacobian/Hessian and QVectorFunc/QGrad. If that is not the case, the user will have to go and manually change the main code. 

Let me know if you have any other suggestions for the naming.

Reply:
I would add this description in the code. It really helps clear things up.

Reply:
Documentation added.